src/profile-logic/marker-data.js

```diff
@@ -388,23 +388,53 @@ export function correlateIPCMarkers(threads: Thread[]): IPCMarkerCorrelations {
 
   const correlations = new IPCMarkerCorrelations();
   for (const markers of markersByKey.values()) {
-    const sendTid = markers[0] ? markers[0].tid : undefined;
-    const recvTid = markers[4] ? markers[4].tid : undefined;
+    const startEndpointMarker = markers[0];
+    const endEndpointMarker = markers[4];
+    const sendTid = startEndpointMarker ? startEndpointMarker.tid : undefined;
+    const recvTid = endEndpointMarker ? endEndpointMarker.tid : undefined;
     const sharedData: IPCSharedData = {
-      startTime: markers[0] ? markers[0].data.startTime : undefined,
+      startTime: startEndpointMarker
+        ? startEndpointMarker.data.startTime
+        : undefined,
       sendStartTime: markers[1] ? markers[1].data.startTime : undefined,
       sendEndTime: markers[2] ? markers[2].data.startTime : undefined,
       recvEndTime: markers[3] ? markers[3].data.startTime : undefined,
-      endTime: markers[4] ? markers[4].data.startTime : undefined,
+      endTime: endEndpointMarker ? endEndpointMarker.data.startTime : undefined,
       sendTid,
       recvTid,
       sendThreadName: formatThreadName(sendTid),
       recvThreadName: formatThreadName(recvTid),
     };
 
-    for (const m of markers) {
-      if (m !== undefined) {
-        correlations.set(m.tid, m.index, sharedData);
+    const addedThreadIds = new Set();
+    if (startEndpointMarker) {
+      addedThreadIds.add(startEndpointMarker.tid);
+      correlations.set(
+        startEndpointMarker.tid,
+        startEndpointMarker.index,
+        sharedData
+      );
+    }
+    if (endEndpointMarker) {
+      addedThreadIds.add(endEndpointMarker.tid);
+      correlations.set(
+        endEndpointMarker.tid,
+        endEndpointMarker.index,
+        sharedData
+      );
+    }
+
+    if (!startEndpointMarker || !endEndpointMarker) {
+      // If both markers from sender and receiver threads are present, we can
+      // only add the markers of them to the correlations. That way, the middle
+      // markers from IPC I/O threads will not be visible in the main thread and
+      // make the main thread crowded. It's important to add all the threads to
+      // the marker.
```

Comment:
If we have both `startEndpointMarker` and `endEndpointMarker` markers, we won't add the IPC I/O markers (the 3 middle ones) to the correlations. This is to reduce the noise in the main thread. Because, when we have the first and last markers, we can directly show them in their correct threads. We only fall back into showing them in the main thread if the IPC message doesn't include the start or end markers (therefore are incomplete). I will explain a bit more in the comment.